File: LeetCode/Python/Medium/159-Longest-Substring-with-At-Most-Two-Distinct-Characters/solution.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        substring=""

        n=len(s)

        hashmap=dict()

        k=2

        left=0

        max_length=0

        substring_start=-1

        substring_end=-1

        for right in range(0, n):
            letter=s[right]

            if letter not in hashmap.keys():
                hashmap[letter]=1
            else:
                hashmap[letter]+=1

            if len(hashmap) <= k:
                substring_length=(right - left) + 1

                if substring_length > max_length:
                    max_length=substring_length

                    substring_start=left

                    substring_end=right
            else:
                while left < right and len(hashmap) > k:
                    if s[left] in hashmap:
                        hashmap[s[left]]-=1

                        if hashmap[s[left]]==0:
                            del hashmap[s[left]]

                    left+=1

        if substring_start!=-1 and substring_end!=-1:
            substring=s[substring_start : substring_end + 1]

        #print(substring)

        return max_length

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        substring=""

        n=len(s)

        hashmap=[0 for _ in range(0, 123)]

        k=2

        left=0

        max_length=0

        substring_start=-1

        substring_end=-1

        for right in range(0, n):
            letter=s[right]

            ascii=ord(letter)

            hashmap[ascii]+=1

            while left < right and self.get_number_of_unique_letters(hashmap) > k:
                left_index=ord(s[left])
                
                if hashmap[left_index] > 0:
                    hashmap[left_index]-=1

                left+=1

            substring_length=(right - left) + 1

            if substring_length > max_length:
                max_length=substring_length

                substring_start=left

                substring_end=right

        if substring_start!=-1 and substring_end!=-1:
            substring=s[substring_start : substring_end + 1]

        #print(substring)

        return max_length

    def get_number_of_unique_letters(self, hashmap: List[int]) -> int:
        unique_letters=0

        for index in range(0, 123):
            if hashmap[index] > 0:
                unique_letters+=1

        return unique_letters

File: LeetCode/Rust/Medium/159-Longest-Substring-with-At-Most-Two-Distinct-Characters/solution.rs
use std::collections::HashMap;

impl Solution
{
    pub fn length_of_longest_substring_two_distinct(s: String) -> i32
    {
        let result: i32=0;

        let n: usize=s.len();

        let k: usize=2;

        let mut hashmap: HashMap<char, u32>=HashMap::new();

        let characters: Vec<char>=s.chars().collect::<Vec<char>>();

        let mut substring_start: i32=-1;

        let mut substring_end: i32=-1;

        let mut max_length: usize=0;

        let mut left: usize=0;

        for right in 0..n
        {
            let letter: char=characters[right];

            hashmap.entry(letter).and_modify(|frequency| *frequency+=1).or_insert(1);

            while left < right && hashmap.len() > k
            {
                if let Some(frequency)=hashmap.get_mut(&characters[left])
                {
                    *frequency-=1;

                    if *frequency==0
                    {
                        hashmap.remove(&characters[left]);
                    }
                }

                left+=1;
            }

            let substring_length: usize=(right - left) + 1;

            if substring_length > max_length
            {
                max_length=substring_length;

                substring_start=left as i32;

                substring_end=right as i32;
            }
        }

        //let substring: &str=&s[substring_start as usize..=substring_end as usize];

        //println!("substring: {}", substring);

        max_length as i32
    }
}

impl Solution
{
    pub fn length_of_longest_substring_two_distinct(s: String) -> i32
    {
        let n: usize=s.len();

        let k: usize=2;

        let mut hashmap: Vec<u32>=vec![0; 123];

        let characters: Vec<char>=s.chars().collect::<Vec<char>>();

        let mut substring_start: i32=-1;

        let mut substring_end: i32=-1;

        let mut max_length: usize=0;

        let mut left: usize=0;

        for right in 0..n
        {
            let letter: char=characters[right];

            let ascii: usize=letter as usize;

            hashmap[ascii]+=1;

            while left < right && Self::get_number_of_unique_letters(&hashmap) > k
            {
                let left_index: usize=characters[left] as usize;

                if hashmap[left_index] > 0
                {
                    hashmap[left_index]-=1;
                }

                left+=1;
            }

            let substring_length: usize=(right - left) + 1;

            if substring_length > max_length
            {
                max_length=substring_length;

                substring_start=left as i32;

                substring_end=right as i32;
            }
        }

        //let substring: &str=&s[substring_start as usize..=substring_end as usize];

        //println!("substring: {}", substring);

        max_length as i32
    }

    pub fn get_number_of_unique_letters(hashmap: &Vec<u32>) -> usize
    {
        let mut unique_letters: usize=0;

        for index in 0..123
        {
            if hashmap[index] > 0
            {
                unique_letters+=1;
            }
        }

        unique_letters
    }
}